Implement input options

ork821 opened the pull request:
Partial implement of #51

orottier commented on src/media/mic.rs:
```diff
@@ -107,7 +109,7 @@ impl Microphone {
 impl Default for Microphone {
     fn default() -> Self {
-        Self::new()
+        Self::new(MicrophoneInitOptions {sample_rate: None})
```
I would prefer
`Self::new(MicrophoneInitOptions::default())`

orottier replied:
I would prefer
`Self::new(MicrophoneInitOptions::default())`

orottier commented on src/media/mic.rs:
```diff
@@ -142,6 +144,15 @@ impl Iterator for Microphone {
+///
+/// Field is optional and will be set to hardcoded value.
+#[derive(Clone, Debug, Default)]
+pub struct MicrophoneInitOptions {
```
Maybe the name `MicrophoneOptions` would be more consistent with `AudioContextOptions`

orottier commented on src/io.rs:
```diff
@@ -403,10 +403,17 @@ pub fn build_input() -> (Stream, StreamConfig, Receiver<AudioBuffer>) {
+        Some(sample_rate)=> supported_configs_range
+            .next()
+            .expect("no supported config?!")
+            .with_sample_rate(CpalSampleRate(sample_rate)),
```
Requesting a specific sample rate should be done after determining the fallback config.
The fallback config is put in the variable `default_config` and should only contain values that were actually supplied by cpal, so they should never fail.
Could you try to set the requested sample rate only on the `config` settings?

orottier reviewed (changes requested):
Hi @ork821, thanks for contributing! Could you have a look at my comments?
Are you planning to add `latency_hint` too in this PR?

ork821 commented:
> Hi @ork821, thanks for contributing! Could you have a look at my comments? Also the CI checks fail, could you run `cargo fmt` and commit those changes too? Are you planning to add `latency_hint` too in this PR?

orottier commented on src/io.rs:
```diff
@@ -425,6 +427,10 @@ pub fn build_input() -> (Stream, StreamConfig, Receiver<AudioBuffer>) {
 
     let mut config: StreamConfig = supported_config.into();
     config.buffer_size = cpal::BufferSize::Fixed(input_buffer_size);
+    if options.sample_rate.is_some() {
```
This can be written better as:

```rust
if let Some(sample_rate) = options.sample_rate {
     // do stuff with sample_rate
}
```

This avoids the `panic` path of the `unwrap` call. See also https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html

orottier commented:
This is going in the right direction :)

I left one nitpick on your code, please have a look.

Also, the build is still failing, you can check it locally with `cargo clippy --all-targets`
It says: `error: this argument is passed by value, but not consumed in the function body` which I think is a false positive and can be ignored with `#[allow(clippy::needless_pass_by_value)]` similar to https://github.com/orottier/web-audio-api-rs/blob/main/src/context/online.rs#L103

After that we can merge and leave the latency for another time

orottier commented:
If you'd like I can get you started with some other issues, maybe #44 or #101

ork821 commented:
> This is going in the right direction :)
> 
> I left one nitpick on your code, please have a look.

I fixed this. Thank you for your help. This is very important for my skills.

> Also, the build is still failing, you can check it locally with `cargo clippy --all-targets` It says: `error: this argument is passed by value, but not consumed in the function body` which I think is a false positive and can be ignored with `#[allow(clippy::needless_pass_by_value)]` similar to https://github.com/orottier/web-audio-api-rs/blob/main/src/context/online.rs#L103

Also fixed this. `cargo build`, `cargo test` and `cargo clippy --all-targets` runs without warnings.

As for other issues, maybe I would try #101. Or will find another issue to fix :)

orottier commented:
Great. You forgot about `cargo fmt` though, but I fixed it for you + an unrelated fix because rust 1.60 was released yesterday.

orottier commented:
@ork821 Since we are nearing a v1 release we need to take care of API stability.
Hence I added the latency_hint option for the mic too (but it is left unimplemented). We can fix this within the v1 version
Please have a look at my latest commit :)

ork821 commented:
> @ork821 Since we are nearing a v1 release we need to take care of API stability. Hence I added the latency_hint option for the mic too (but it is left unimplemented). We can fix this within the v1 version Please have a look at my latest commit :)

Hi, @orottier. Thank you for fixing my small `cargo fmt` fault. 

I am not sure that usage same struct `AudioContextOptions` is good desition. 
For example, according to https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/AudioContext we can have only `latencyHint` and `sampleRate`. But according to https://w3c.github.io/mediacapture-main/#dom-mediatrackconstraintset we can have `echoCancellation`, `channelCount`, `deviceId` (and so on) as input device options. If for now it's ok, we can merge.

orottier commented:
That's a valid point, however I would prefer to keep it this way, because either way we need breaking changes when we add the features that you mention to `AudioInputOptions`.
https://doc.rust-lang.org/cargo/reference/semver.html#struct-add-public-field-when-no-private

Let's op for a new constructor in the future
`fn new_with_config(audio_opts: AudioContextOptions, mic_opts: MicrophoneOptions) -> Microphone`
We can add all mic specific fields to `MicrophoneOptions`. The new constructor is not a breaking change so can be added within the v1 version. This gives us a bit of time to flesh out the complete featureset around mic input

ork821 commented:
> That's a valid point, however I would prefer to keep it this way, because either way we need breaking changes when we add the features that you mention to `AudioInputOptions`. https://doc.rust-lang.org/cargo/reference/semver.html#struct-add-public-field-when-no-private
> 
> Let's op for a new constructor in the future `fn new_with_config(audio_opts: AudioContextOptions, mic_opts: MicrophoneOptions) -> Microphone` We can add all mic specific fields to `MicrophoneOptions`. The new constructor is not a breaking change so can be added within the v1 version. This gives us a bit of time to flesh out the complete featureset around mic input

Yes, you are actually right about breaking changes. It would be better to implement mentioned method later if we need it. I think you can merge if everything is okay.
